src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt

```diff
@@ -58,18 +56,31 @@ internal class LoadTest(
     private val load = behavior.load
 
     fun run() {
+        val users = generateUsers()
         logger.info("Holding for ${load.hold}.")
         Thread.sleep(load.hold.toMillis())
         workspace.toFile().ensureDirectory()
         setUpJira()
-        applyLoad()
+        applyLoad(users)
         val nodesDump = workspace.resolve("nodes.csv")
         nodesDump.toFile().bufferedWriter().use {
             nodeCounter.dump(it)
         }
         logger.debug("Dumped node's counts to $nodesDump")
     }
 
+    private fun generateUsers(): List<User> {
+        val pool = Executors.newCachedThreadPool { runnable ->
+            Thread(runnable, "user-generation-${runnable.hashCode()}")
+        }
+        val deadline = now() + behavior.maxOverhead
+        val users = (1..load.virtualUsers)
+            .map { pool.submit(Callable { userGenerator.generateUser(options) }) }
+            .map { it.finishBy(deadline, logger) }
+        pool.shutdownNow()
```

Comment:
What do you think about `shutdownNow ` in `finally` block?

Reply:
Haha 😀 you got me there
I'll `finally` do [it](https://bitbucket.org/atlassian/jira-performance-tests/pull-requests/76/jperf-123-a-sample-test-for-running-dc/diff#comment-78922842) 😉

Reply:
Raised and assigned https://ecosystem.atlassian.net/browse/JPERF-546

Reply:
I wonder if that’s why my assertions inside Actions were breaking tests.

Reply:
VU threads are fire-and-forget, so the unhappy-path-shutdown wouldn't matter. Any `Error` thrown from an `Action` will kill the VU and result in fewer measurements.

Reply:
I pushed in `AbruptExecutorService` usage